hazelcast/src/main/java/com/hazelcast/function/FunctionsImpl.java

```diff
@@ -0,0 +1,133 @@
+            List<Permission> beforeP = before.permissions();
+            List<Permission> afterP = after.permissions();
+            return sumPermissions(afterP, beforeP);
+        }
+
+        @Nullable
+        private static List<Permission> sumPermissions(@Nullable List<Permission> afterP, @Nullable List<Permission> beforeP) {
+            if (afterP == null || afterP.isEmpty()) {
+                return beforeP;
+            } else if (beforeP == null || beforeP.isEmpty()) {
+                return afterP;
+            } else {
+                List<Permission> permissions = new ArrayList<>(afterP.size() + beforeP.size());
+                permissions.addAll(beforeP);
+                permissions.addAll(afterP);
+                return permissions;
+            }
+        }
+    }
+
+    @SerializableByConvention
+    static final class ComposedFunctionEx<V, T, R>
+            extends ComposedSecuredFunction<FunctionEx<? super V, ? extends T>, FunctionEx<? super T, ? extends R>>
+            implements FunctionEx<V, R> {
+
+        ComposedFunctionEx(@Nonnull FunctionEx<? super V, ? extends T> before,
+                                  @Nonnull FunctionEx<? super T, ? extends R> after) {
+            super(before, after);
+        }
+
+        @Override
+        public R applyEx(V v) throws Exception {
+            return after.applyEx(before.applyEx(v));
+        }
+    }
+
+    @SerializableByConvention
+    static final class ComposedBiFunctionEx<U, V, T, R>
+            extends ComposedSecuredFunction<BiFunctionEx<? super U, ? super V, ? extends T>, FunctionEx<? super T, ? extends R>>
+            implements BiFunctionEx<U, V, R> {
+
+        ComposedBiFunctionEx(@Nonnull BiFunctionEx<? super U, ? super V, ? extends T> before,
+                                  @Nonnull FunctionEx<? super T, ? extends R> after) {
+            super(before, after);
+        }
+
+        @Override
+        public R applyEx(U t, V u) throws Exception {
+            return after.applyEx(before.applyEx(t, u));
+        }
+    }
+
+    @SerializableByConvention
+    static final class ComposedConsumerEx<T>
+            extends ComposedSecuredFunction<ConsumerEx<? super T>, ConsumerEx<? super T>>
+            implements ConsumerEx<T> {
+
+        ComposedConsumerEx(@Nonnull ConsumerEx<? super T> before,
+                                    @Nonnull ConsumerEx<? super T> after) {
+            super(before, after);
+        }
+
+        @Override
+        public void acceptEx(T t) throws Exception {
+            before.acceptEx(t);
+            after.acceptEx(t);
+        }
+    }
+
+    @SerializableByConvention
+    static final class ComposedSupplierEx<T, R>
+            extends ComposedSecuredFunction<SupplierEx<? extends T>, FunctionEx<? super T, ? extends R>>
+            implements SupplierEx<R> {
+
+        ComposedSupplierEx(SupplierEx<? extends T> wrapped, FunctionEx<? super T, ? extends R> wrapperSupplier) {
```

Comment:
I'm confused about the argument names. Wouldn't naming the arguments `wrappedSupplier` and `wrapper` make more sense?  (or `wrapped` and `supplierWrapper`)

Reply:
I think I copied it from `WrappingProcessorSupplier`, but agree that it is hard to understand. Changed to ordinary before/after.
[ebffe21](https://github.com/hazelcast/hazelcast/pull/25966/commits/ebffe2101dcf43fb1b64f72086232626d3ac6471)